### PR TITLE
Gql error messages

### PR DESCRIPTION
**Pull request by asmab89**

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt [hunk 1]
```diff
@@ -461,48 +461,73 @@ object GeneUtils {
...
+    fun repairBooleanSelection(obj: ObjectGene): Boolean {
 
         if (obj.fields.isEmpty()
             || obj.fields.count { it !is OptionalGene && it !is BooleanGene && it !is TupleGene } > 0
         ) {
             throw IllegalArgumentException("There should be at least 1 field, and they must be all optional or boolean or tuple")
         }
 
+        /*
+        * Filtering functional genes among all genes.
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:473]: what is a `functional` gene?

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt [hunk 2]
```diff
@@ -461,48 +461,73 @@ object GeneUtils {
...
             selected.forEach {
                 if ((it is OptionalGene && it.gene is ObjectGene && (it.gene !is CycleObjectGene || it.gene !is LimitObjectGene))
                 ) {
-                    repairBooleanSelection(it.gene)
-                } else if ( //looking into objects inside a tuple
-                    isTupleOptionalObjetNotCycleNotLimit(it)) {
-                    repairBooleanSelection(((it as TupleGene).elements.last() as OptionalGene).gene as ObjectGene)
-                }
+                    if (!repairBooleanSelection(it.gene)) {
+                        it.isActive = false
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:495]: what happen if all cases inside `selected` endups being put to `false`?

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt [hunk 3]
```diff
@@ -461,48 +461,73 @@ object GeneUtils {
...
         }
 
+        /*
+        * we have found some functional genes.
+        * we need to look into them to find functional genes as well
+        * */
         if (selected.isNotEmpty()) {
             //it is fine, but we still need to make sure selected objects are fine
             selected.forEach {
                 if ((it is OptionalGene && it.gene is ObjectGene && (it.gene !is CycleObjectGene || it.gene !is LimitObjectGene))
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:492]: `CycleObjectGene` and `LimitObjectGene` no longer extends `ObjectGene`, so these checks seem redundant

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt [hunk 4]
```diff
@@ -461,48 +461,73 @@ object GeneUtils {
...
-                }
+                    if (!repairBooleanSelection(it.gene)) {
+                        it.isActive = false
+                    }
+                } else  //looking into objects inside a tuple
+                    if (isTupleOptionalObjetNotCycleNotLimit(it)) {
+                        if (!repairBooleanSelection(((it as TupleGene).elements.last() as OptionalGene).gene as ObjectGene)) {
+                            (it.elements.last() as OptionalGene).isActive = false
+                        }
+                    } else return false
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:502]: explain the `return false` here

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt [hunk 5]
```diff
@@ -461,48 +461,73 @@ object GeneUtils {
...
             //must select at least one
             val candidates = obj.fields.filter {
                 (it is OptionalGene && it.selectable) || it is BooleanGene ||
                         (it is TupleGene && it.lastElementTreatedSpecially && isLastCandidate(it))
             }
-            assert(candidates.isNotEmpty())
+
+
+            /*
+            * What if we do not have candidates? we could not recycle any gene !.
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:517]: what does it mean to `recycle` a gene?

**Review by arcuri82 — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt (hunk 1 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt]: resolved

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt (hunk 2 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:510]: resolved

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt (hunk 3 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt]: Removed

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt (hunk 4 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt]: removed

File: core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt (hunk 5 above)
Comment (asmab89) [reply core/src/main/kotlin/org/evomaster/core/search/gene/utils/GeneUtils.kt:537]: resolved

**Pull request merged**